src/main/java/com/commercetools/sync/categories/helpers/CategoryReferenceResolver.java

```diff
@@ -0,0 +1,116 @@
+package com.commercetools.sync.categories.helpers;
+
+
+import com.commercetools.sync.categories.CategorySyncOptions;
+import com.commercetools.sync.commons.exceptions.ReferenceResolutionException;
+import com.commercetools.sync.commons.helpers.BaseReferenceResolver;
+import com.commercetools.sync.services.CategoryService;
+import com.commercetools.sync.services.TypeService;
+import io.sphere.sdk.categories.Category;
+import io.sphere.sdk.categories.CategoryDraft;
+import io.sphere.sdk.categories.CategoryDraftBuilder;
+import io.sphere.sdk.models.Reference;
+import io.sphere.sdk.types.CustomFieldsDraft;
+import io.sphere.sdk.utils.CompletableFutureUtils;
+import org.apache.commons.lang3.StringUtils;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+public final class CategoryReferenceResolver extends BaseReferenceResolver<CategoryDraft, CategorySyncOptions> {
+    private CategoryService categoryService;
+
+    public CategoryReferenceResolver(@Nonnull final CategorySyncOptions options,
+                                     @Nonnull final TypeService typeService,
+                                     @Nonnull final CategoryService categoryService) {
+        super(options, typeService);
+        this.categoryService = categoryService;
+    }
+
+    @Override
+    @Nonnull
+    public CompletionStage<CategoryDraft> resolveCustomTypeReference(@Nonnull final CategoryDraft categoryDraft) {
+        final CustomFieldsDraft custom = categoryDraft.getCustom();
+        if (custom != null) {
+            return getCustomTypeId(custom)
+                .thenApply(resolvedTypeIdOptional ->
+                    resolvedTypeIdOptional.filter(StringUtils::isNotBlank)
+                                          .map(resolvedTypeId ->
+                                              CategoryDraftBuilder.of(categoryDraft)
+                                                                  .custom(CustomFieldsDraft.ofTypeIdAndJson(
+                                                                      resolvedTypeId, custom.getFields()))
+                                                                  .build())
+                                          .orElseGet(() -> CategoryDraftBuilder.of(categoryDraft).build()));
```

Comment:
because here the map() returns an optional containing a `CategoryDraftDsl` so the orElseGet should return the same type and `categoryDraft` is passed as a `CategoryDraft` not `CategoryDraftDsl` (which actually implements it). If there is a better way? please suggest :)